src/lib.rs

```
@@ -1,82 +1,183 @@
 #![no_std]
 
+mod impls;
+
 extern crate alloc;
 
 use alloc::borrow::Cow;
-use auto_impl::auto_impl;
 
 /// Merkle root alias type
 pub type MerkleRoot = [u8; 32];
 
-/// Base map trait for Fuel infrastructure
-///
-/// Generics:
-///
-/// - K: Key that maps to a value
-/// - V: Stored value
-#[auto_impl(&mut)]
-pub trait Storage<K, V>
-where
-    V: Clone,
-{
-    /// Error implementation of the storage functions
+/// Mappable type with `Key` and `Value`.
+pub trait Mappable {
+    /// The type of the value's key.
+    type Key;
+    /// The value type is used while setting the value to the storage. In most cases, it is the same
+    /// as `Self::GetValue`, but it is without restriction and can be used for performance
+    /// optimizations.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use core::marker::PhantomData;
+    /// use fuel_storage::Mappable;
+    /// pub struct Contract<'a>(PhantomData<&'a ()>);
+    ///
+    /// impl<'a> Mappable for Contract<'a> {
+    ///     type Key = &'a [u8; 32];
+    ///     /// It is optimized to use slice instead of vector.
+    ///     type SetValue = [u8];
+    ///     type GetValue = Vec<u8>;
+    /// }
+    /// ```
+    type SetValue: ?Sized;
+    /// The value type is used while getting the value from the storage.
+    type GetValue: Clone;
+}
+
+/// Trait describes used errors during work with `Storage` for the `Type`.
+pub trait StorageError<Type: Mappable> {
```

Do we expect specific behavior for the error? Also, why should it be tied to a concrete key/value implementation instead of the map itself?

Looks to me that most errors will be in fact result of the map implementation instead of specific derivations of the stored type. Example: a map might fail to fetch a hash from the disk, but shouldn't serialize it incorrectly so it can't be deserialized later.

---

The `Error` can be part of `Mappable`, but it will be not so customizable. The `Error` in the `Mappable` means that different implementations of the `Mappable` should return the same error. You can implement `Storage<Mappable>` for the `RocksDB` and here you may return `RockDBError`, but for the `MemoryDB` you may want to return another error. The `Mappable` type is the same but errors for different implementations are different=)

---

This seems like it would create a lot of extra boilerplate if we have to implement this trait for each mappable x database type combo. Why not just make Error a generic associated type of `StorageInspect`/`StorageMutate`?

---

I tried to make`StorageInspect` and `StorageMutate` independent, but I also like the following dependence:

```rust
pub trait StorageInspect<Type: Mappable> {
    type Error;
    ...
}

pub trait StorageMutate<Type: Mappable>: StorageInspect<Type> {
    ...
}
```

If you prefer that variant, I'm okay to go with it=) (You can like my comment and I will commit as above)

---

*liked*

---

@vlopes11 I think this is resolved now since this trait no longer exists.